File: kotlin-mp-collections-jvm/src/main/java/edu/alex/RefMap.java
package edu.alex;

import static edu.alex.ObjectReference.*;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class RefMap<K, V> implements Map<K, V> {

    private final static <K, V> Entry<ObjectReference<K>, V> wrapEntry(Entry<K, V> e) {
        return new AbstractMap.SimpleEntry<>(wrap(e.getKey()), e.getValue());
    }

    private final static <K, V> Entry<K, V> unwrapEntry(Entry<ObjectReference<K>, V> e) {
        return new AbstractMap.SimpleEntry<>(e.getKey().get(), e.getValue());
    }

    private final static <K, V> Stream<Entry<ObjectReference<K>, V>> toWrappedStream(Collection<?>  c) {
        return c.stream()
                .filter(e -> e instanceof Entry)
                .map(e -> wrapEntry((Entry)e));
    }


    private final Map<ObjectReference<K>, V> backingMap = new LinkedHashMap<>();

    @Override
    public int size() {
        return backingMap.size();
    }

    @Override
    public boolean isEmpty() {
        return backingMap.isEmpty();
    }

    @Override
    public V remove(Object key) {
        return backingMap.remove(wrap(key));
    }

    @Override
    public V get(Object key) {
        return backingMap.get(wrap(key));
    }

    @Override
    public boolean containsKey(Object key) {
        return backingMap.containsKey(wrap(key));
    }

    @Override
    public boolean containsValue(Object value) {
        return backingMap.containsValue(value);
    }

    @Override
    public V put(K key, V value) {
        return backingMap.put(wrap(key), value);
    }

    @Override
    public void putAll(Map<? extends K, ? extends V> m) {
        m.forEach((key, value) -> backingMap.put(wrap(key), value));
    }

    @Override
    public void clear() {
        backingMap.clear();
    }

    @Override
    public Collection<V> values() {
        return backingMap.values();
    }

    @Override
    public Set<K> keySet() {
        return new RefSet(backingMap.keySet());
    }

    @Override
    @SuppressWarnings("unchecked")
    public Set<Entry<K, V>> entrySet() {
        final Set<Entry<ObjectReference<K>, V>> backingSet = backingMap.entrySet();

        return new Set<Entry<K, V>>() {
            @Override
            public int size() {
                return backingSet.size();
            }

            @Override
            public boolean isEmpty() {
                return backingSet.isEmpty();
            }

            @Override
            public boolean contains(Object o) {
                if(o instanceof Entry) {
                    final Entry<K, V> entry = (Entry) o;
                    return backingSet.contains(wrapEntry(entry));
                }
                return false;
            }

            @Override
            public Iterator<Entry<K, V>> iterator() {
                final Iterator<Entry<ObjectReference<K>, V>> it = backingSet.iterator();

                return new Iterator<Entry<K, V>>() {
                    @Override
                    public boolean hasNext() {
                        return it.hasNext();
                    }

                    @Override
                    public Entry<K, V> next() {
                        return unwrapEntry(it.next());
                    }

                    @Override
                    public void remove() {
                        it.remove();
                    }
                };
            }

            @Override
            public Object[] toArray() {
                return backingSet.stream()
                        .map(RefMap::unwrapEntry).toArray();
            }

            @Override
            public <T> T[] toArray(T[] a) {
                return backingSet.stream()
                        .map(RefMap::unwrapEntry)
                        .collect(Collectors.toList()).toArray(a);
            }

            @Override
            public boolean add(Entry<K, V> kvEntry) {
                return backingSet.add(wrapEntry(kvEntry));
            }

            @Override
            public boolean remove(Object o) {
                if(o instanceof Entry) {
                    final Entry<K, V> entry = (Entry) o;
                    return backingSet.remove(wrapEntry(entry));
                }
                return false;
            }

            @Override
            public boolean containsAll(Collection<?> c) {
                return c.stream()
                        .allMatch(this::contains);
            }

            @Override
            public boolean addAll(Collection<? extends Entry<K, V>> c) {
                return c.stream()
                        .map(RefMap::wrapEntry)
                        .reduce(false, (anyChange, o) -> backingSet.add(o) || anyChange, (a, b) -> a || b);
            }

            @Override
            public boolean retainAll(Collection<?> c) {
                return backingSet.retainAll(toWrappedStream(c)
                        .collect(Collectors.toSet()));
            }

            @Override
            public boolean removeAll(Collection<?> c) {
                return toWrappedStream(c)
                        .reduce(false, (anyChange, e) -> backingSet.remove(e) || anyChange, (a, b) -> a || b);
            }

            @Override
            public void clear() {
                backingSet.clear();
            }
        };
    }
}

File: kotlin-mp-collections-jvm/src/main/kotlin/edu/alex/RefMap.kt
package edu.alex

import java.util.LinkedHashMap
import edu.alex.ObjectReference.wrap

class RefMap<K, V>: MutableMap<K, V> {

    private companion object {
        fun <K, V> wrapEntry(e: MutableMap.MutableEntry<K, V>): MutableMap.MutableEntry<ObjectReference<K>, V> {
            return object : MutableMap.MutableEntry<ObjectReference<K>, V> {
                override val key = wrap<K>(e.key)
                override val value: V get() = e.value
                override fun setValue(value: V): V = e.setValue(value)
            }
        }
    }

    private fun unwrapEntry(e: MutableMap.MutableEntry<ObjectReference<K>, V>): MutableMap.MutableEntry<K, V> {
        var lastValue = e.value
        return object : MutableMap.MutableEntry<K, V> {
            override val key = e.key.get()
            override val value: V get() = lastValue

            override fun setValue(value: V): V {
                val tmp = lastValue
                lastValue = value
                backingMap[e.key] = value
                return tmp
            }
        }
    }


    private val backingMap = LinkedHashMap<ObjectReference<K>, V>()
    private val entrySet = backingMap.entries

    override val size: Int get() = backingMap.size

    override fun clear() = backingMap.clear()

    override fun containsKey(key: K) = backingMap.containsKey(wrap(key))

    override fun containsValue(value: V) = backingMap.containsValue(value)

    override fun get(key: K): V? = backingMap[wrap(key)]

    override fun isEmpty() = backingMap.isEmpty()

    override fun put(key: K, value: V): V? = backingMap.put(wrap(key), value)

    override fun putAll(from: Map<out K, V>) = from.forEach { (key, value) -> put(key, value) }

    override fun remove(key: K): V? = backingMap.remove(wrap(key))

    override val values: MutableCollection<V> get() = backingMap.values

    override val keys: MutableSet<K> get() = RefSet(backingMap.keys)

    override val entries: MutableSet<MutableMap.MutableEntry<K, V>> get() =
        object : MutableSet<MutableMap.MutableEntry<K, V>> {
            override val size: Int get() = entrySet.size

            override fun add(element: MutableMap.MutableEntry<K, V>) = entrySet.add(wrapEntry(element))

            override fun remove(element: MutableMap.MutableEntry<K, V>) = entrySet.remove(wrapEntry(element))

            override fun addAll(elements: Collection<MutableMap.MutableEntry<K, V>>) =
                    elements.fold(false) { anyChange, e -> add(e) || anyChange }

            override fun clear() = entrySet.clear()

            override fun iterator(): MutableIterator<MutableMap.MutableEntry<K, V>> {
                var it = entrySet.iterator()
                return object: MutableIterator<MutableMap.MutableEntry<K, V>> {
                    override fun hasNext() = it.hasNext()
                    override fun next() = unwrapEntry(it.next())
                    override fun remove() = it.remove()
                }
            }

            override fun removeAll(elements: Collection<MutableMap.MutableEntry<K, V>>) =
                    elements.fold(false) { anyChange, e -> remove(e) || anyChange }

            override fun retainAll(elements: Collection<MutableMap.MutableEntry<K, V>>) =
                    entrySet.retainAll( elements.map { wrapEntry(it) })

            override fun contains(element: MutableMap.MutableEntry<K, V>) = entrySet.contains(wrapEntry(element))

            override fun containsAll(elements: Collection<MutableMap.MutableEntry<K, V>>) = elements.all { contains(it) }

            override fun isEmpty() = size == 0
        }

    fun set(key: K, value: V): RefMap<K, V> {
        put(key, value)
        return this
    }

    fun delete(key: K?): RefMap<K, V> {
        remove(key)
        return this
    }
}